R/sampler-hmc.R

```diff
@@ -0,0 +1,222 @@
+##' Create a Hamiltonian Monte Carlo sampler, implemented using the
+##' leapfrog algorithm.
+##'
+##' @title Create HMC
+##'
+##' @param epsilon The step size of the HMC steps
+##'
+##' @param n_integration_steps The number of HMC steps per step
+##'
+##' @param vcv A variance-covariance matrix for the momentum vector.
+##'   The default uses an identity matrix.
+##'
+##' @param debug Logical, indicating if we should save all
+##'   intermediate points and their gradients.  This will add a vector
+##'   "history" to the details after the integration.  This *will*
+##'   slow things down though as we accumulate the history
+##'   inefficiently.
+##'
+##' @return A `mcstate_sampler` object, which can be used with
+##'   [mcstate_sample]
+##'
+##' @export
+mcstate_sampler_hmc <- function(epsilon = 0.015, n_integration_steps = 10,
+                                vcv = NULL, debug = FALSE) {
+  internal <- new.env()
+
+  if (!is.null(vcv)) {
+    check_vcv(vcv, call = environment())
+  }
+
+  initialise <- function(state, model, rng) {
+    internal$transform <- hmc_transform(model$domain)
+    n_pars <- length(model$parameters)
+    if (is.null(vcv)) {
+      internal$sample_momentum <- function(rng) rng$random_normal(n_pars)
+    } else {
+      if (n_pars != nrow(vcv)) {
+        cli::cli_abort(
+          "Incompatible length parameters ({n_pars}) and vcv ({nrow(vcv)})")
+      }
+      internal$sample_momentum <- make_rmvnorm(vcv, centred = TRUE)
+    }
+    if (debug) {
+      internal$history <- list()
+    }
+  }
+
+  step <- function(state, model, rng) {
+    ## Just a helper for now
+    compute_gradient <- function(x) {
+      internal$transform$deriv(x) * model$gradient(x)
+    }
+
+    theta <- internal$transform$model2rn(state$pars)
+    pars_next <- state$pars
+    theta_next <- theta
+
+    if (debug) {
+      history <- matrix(NA_real_, n_integration_steps + 1, length(theta))
+      history[1, ] <- pars_next
+    }
+
+    v <- internal$sample_momentum(rng)
+
+    ## Make a half step for momentum at the beginning
+    v_next <- v + epsilon * compute_gradient(pars_next) / 2
+
+    for (i in seq_len(n_integration_steps)) {
+      ## Make a full step for the position
+      theta_next <- theta_next + epsilon * v_next
+      pars_next <- internal$transform$rn2model(theta_next)
+      ## Make a full step for the momentum, except at end of trajectory
+      if (i != n_integration_steps) {
+        v_next <- v_next + epsilon * compute_gradient(pars_next)
+      }
+      if (debug) {
+        history[i + 1, ] <- pars_next
+      }
+    }
+
+    ## Make a half step for momentum at the end.
+    v_next <- v_next + epsilon * compute_gradient(pars_next) / 2
+
+    ## Negate momentum at end of trajectory to make the proposal symmetric
```

Comment:
The negative change can just be dropped here as vm is independently resampled at each step.